File: frontend/steward-ui/types/customer.ts
/**
 * Customer (CIF) TypeScript types
 */

export type CustomerType = 'INDIVIDUAL' | 'CORPORATE' | 'SYARIAH';
export type Gender = 'MALE' | 'FEMALE';
export type MaritalStatus = 'SINGLE' | 'MARRIED' | 'DIVORCED' | 'WIDOWED';
export type KycStatus = 'PENDING' | 'IN_REVIEW' | 'APPROVED' | 'REJECTED' | 'EXPIRED' | 'FAILED';
export type RiskProfile = 'LOW' | 'MEDIUM' | 'HIGH';
export type CustomerStatus = 'ACTIVE' | 'DORMANT' | 'CLOSED' | 'BLACKLIST' | 'DECEASED' | 'SUSPENDED';

export interface Customer {
  id: string;
  cifNumber: string;
  customerType: CustomerType;
  fullName?: string;
  legalName?: string;
  dateOfBirth?: string;
  placeOfBirth?: string;
  gender?: Gender;
  nationality?: string;
  maritalStatus?: MaritalStatus;
  religion?: string;
  occupation?: string;
  monthlyIncome?: number;
  kycStatus: KycStatus;
  kycExpiryDate?: string;
  riskProfile?: RiskProfile;
  pepStatus: boolean;
  cifStatus: CustomerStatus;
  branchId?: string;
  regionCode?: string;
  tags?: string[];
  customFields?: Record<string, any>;
  createdAt: string;
  createdBy?: string;
  updatedAt: string;
  updatedBy?: string;
  version: number;
}

export interface CustomerSearchFilters {
  cifStatus?: CustomerStatus;
  kycStatus?: KycStatus;
  riskProfile?: RiskProfile;
  branchId?: string;
  customerType?: CustomerType;
  dateOfBirthFrom?: string;
  dateOfBirthTo?: string;
  monthlyIncomeMin?: number;
  monthlyIncomeMax?: number;
  [key: string]: any;
}

export interface CustomerCreatePayload {
  customerType: CustomerType;
  fullName?: string;
  legalName?: string;
  dateOfBirth?: string;
  placeOfBirth?: string;
  gender?: Gender;
  nationality?: string;
  maritalStatus?: MaritalStatus;
  religion?: string;
  occupation?: string;
  monthlyIncome?: number;
  branchId?: string;
  email?: string;
  mobilePhone?: string;
  customFields?: Record<string, any>;
  tags?: string[];
}

export interface CustomerUpdatePayload {
  fullName?: string;
  legalName?: string;
  dateOfBirth?: string;
  placeOfBirth?: string;
  gender?: Gender;
  nationality?: string;
  maritalStatus?: MaritalStatus;
  religion?: string;
  occupation?: string;
  monthlyIncome?: number;
  kycStatus?: KycStatus;
  kycExpiryDate?: string;
  riskProfile?: RiskProfile;
  pepStatus?: boolean;
  cifStatus?: CustomerStatus;
  branchId?: string;
  email?: string;
  mobilePhone?: string;
  customFields?: Record<string, any>;
  tags?: string[];
}

// Status badge color mapping
export const STATUS_BADGE_TYPES: Record<string, string> = {
  // Customer status
  ACTIVE: 'success',
  APPROVED: 'success',
  DORMANT: 'info',
  PENDING: 'warning',
  IN_REVIEW: 'warning',
  CLOSED: 'info',
  EXPIRED: 'info',
  REJECTED: 'danger',
  FAILED: 'danger',
  BLACKLIST: 'danger',
  DECEASED: 'info',
  SUSPENDED: 'warning',
  // Risk
  LOW: 'success',
  MEDIUM: 'warning',
  HIGH: 'danger',
};

// Constants for selects
export const CUSTOMER_TYPES: { value: CustomerType; label: string }[] = [
  { value: 'INDIVIDUAL', label: 'Individual' },
  { value: 'CORPORATE', label: 'Corporate' },
  { value: 'SYARIAH', label: 'Syariah' },
];

export const GENDERS: { value: Gender; label: string }[] = [
  { value: 'MALE', label: 'Laki-laki' },
  { value: 'FEMALE', label: 'Perempuan' },
];

export const MARITAL_STATUSES: { value: MaritalStatus; label: string }[] = [
  { value: 'SINGLE', label: 'Belum Menikah' },
  { value: 'MARRIED', label: 'Menikah' },
  { value: 'DIVORCED', label: 'Cerai' },
  { value: 'WIDOWED', label: 'Duda/Janda' },
];

export const RELIGIONS = [
  'ISLAM',
  'KRISTEN',
  'KATOLIK',
  'HINDU',
  'BUDDHA',
  'KONGHUCU',
  'LAINNYA',
];

export const KYC_STATUSES: { value: KycStatus; label: string; type: string }[] = [
  { value: 'PENDING', label: 'Pending', type: 'warning' },
  { value: 'IN_REVIEW', label: 'In Review', type: 'warning' },
  { value: 'APPROVED', label: 'Approved', type: 'success' },
  { value: 'REJECTED', label: 'Rejected', type: 'danger' },
  { value: 'EXPIRED', label: 'Expired', type: 'info' },
  { value: 'FAILED', label: 'Failed', type: 'danger' },
];

export const RISK_PROFILES: { value: RiskProfile; label: string; type: string }[] = [
  { value: 'LOW', label: 'Low Risk', type: 'success' },
  { value: 'MEDIUM', label: 'Medium Risk', type: 'warning' },
  { value: 'HIGH', label: 'High Risk', type: 'danger' },
];

export const CUSTOMER_STATUSES: { value: CustomerStatus; label: string; type: string }[] = [
  { value: 'ACTIVE', label: 'Active', type: 'success' },
  { value: 'DORMANT', label: 'Dormant', type: 'info' },
  { value: 'CLOSED', label: 'Closed', type: 'info' },
  { value: 'BLACKLIST', label: 'Blacklist', type: 'danger' },
  { value: 'DECEASED', label: 'Deceased', type: 'info' },
  { value: 'SUSPENDED', label: 'Suspended', type: 'warning' },
];

File: backend/customer-service/src/main/java/com/bankxyz/mdm/customer/domain/Customer.java
package com.bankxyz.mdm.customer.domain;

import com.bankxyz.mdm.common.entity.BaseEntity;
import io.hypersistence.utils.hibernate.type.json.JsonBinaryType;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import org.hibernate.annotations.Type;

import java.time.LocalDate;
import java.util.Map;
import java.util.UUID;

/**
 * Customer (CIF - Customer Information File) entity.
 * Core entity that represents a single customer in the system.
 *
 * Customer types:
 * - INDIVIDUAL: Natural person (personal banking)
 * - CORPORATE: Legal entity (business banking)
 * - SYARIAH: Sharia-compliant banking customer
 */
@Entity
@Table(name = "customers", indexes = {
    @Index(name = "idx_customers_cif_number", columnList = "cif_number", unique = true),
    @Index(name = "idx_customers_full_name", columnList = "full_name"),
    @Index(name = "idx_customers_status", columnList = "cif_status"),
    @Index(name = "idx_customers_branch", columnList = "branch_id"),
    @Index(name = "idx_customers_kyc_status", columnList = "kyc_status, kyc_expiry_date"),
    @Index(name = "idx_customers_risk", columnList = "risk_profile"),
    @Index(name = "idx_customers_created", columnList = "created_at DESC")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
public class Customer extends BaseEntity {

    @Column(name = "cif_number", nullable = false, unique = true, length = 30)
    private String cifNumber;

    @Enumerated(EnumType.STRING)
    @Column(name = "customer_type", nullable = false, length = 20)
    private CustomerType customerType;

    // For individuals
    @Column(name = "full_name", length = 200)
    private String fullName;

    @Column(name = "legal_name", length = 300)
    private String legalName;

    @Column(name = "date_of_birth")
    private LocalDate dateOfBirth;

    @Column(name = "place_of_birth", length = 100)
    private String placeOfBirth;

    @Enumerated(EnumType.STRING)
    @Column(name = "gender", length = 10)
    private Gender gender;

    @Column(name = "nationality", length = 3)
    private String nationality;  // ISO 3166-1 alpha-3 (ID, SG, MY, etc)

    @Enumerated(EnumType.STRING)
    @Column(name = "marital_status", length = 20)
    private MaritalStatus maritalStatus;

    @Column(name = "religion", length = 20)
    private String religion;

    @Column(name = "occupation", length = 100)
    private String occupation;

    @Column(name = "monthly_income", precision = 18, scale = 2)
    private java.math.BigDecimal monthlyIncome;

    // KYC
    @Enumerated(EnumType.STRING)
    @Column(name = "kyc_status", nullable = false, length = 20)
    private KycStatus kycStatus;

    @Column(name = "kyc_expiry_date")
    private LocalDate kycExpiryDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "risk_profile", length = 10)
    private RiskProfile riskProfile;

    @Column(name = "pep_status", nullable = false)
    private Boolean pepStatus;  // Politically Exposed Person

    // Status
    @Enumerated(EnumType.STRING)
    @Column(name = "cif_status", nullable = false, length = 20)
    private CustomerStatus cifStatus;

    // Location
    @Column(name = "branch_id", length = 20)
    private String branchId;

    @Column(name = "region_code", length = 20)
    private String regionCode;

    // Metadata
    @Type(JsonBinaryType.class)
    @Column(name = "tags", columnDefinition = "jsonb")
    private String[] tags;  // Custom tags (PREMIUM, VIP, etc)

    @Type(JsonBinaryType.class)
    @Column(name = "custom_fields", columnDefinition = "jsonb")
    private Map<String, Object> customFields;

    @PrePersist
    public void prePersist() {
        if (kycStatus == null) kycStatus = KycStatus.PENDING;
        if (cifStatus == null) cifStatus = CustomerStatus.ACTIVE;
        if (pepStatus == null) pepStatus = false;
    }

    public enum CustomerType { INDIVIDUAL, CORPORATE, SYARIAH }
    public enum Gender { MALE, FEMALE }
    public enum MaritalStatus { SINGLE, MARRIED, DIVORCED, WIDOWED }
    public enum KycStatus { PENDING, IN_REVIEW, APPROVED, REJECTED, EXPIRED, FAILED }
    public enum RiskProfile { LOW, MEDIUM, HIGH }
    public enum CustomerStatus { ACTIVE, DORMANT, CLOSED, BLACKLIST, DECEASED, SUSPENDED }
}
